chainer/graph_optimizations/static_graph_utilities.py

```diff
@@ -0,0 +1,197 @@
+    completely static array allocations if of course not possible if
+    any of the schedule functions return a tuple of dynamically allocated
+    arrays.
+
+    The following optional arguments apply to the wrapped function or method.
+
+    Args:
+        inputs (list of ndarray): An optional keyword argument that
+            supplies all arrays needed as input by the function. If the
+            function needs an array that is used by another function
+            in the static schedule, it must appear in this list.
+        outputs (list of ndarray): An optional keyword argument that
+            supplies all output arrays of this function. These arrays
+            must already have been initialized to the correct shapes
+            and dtypes before the function is called. The function
+            must write its results in-place into these arrays. Any
+            output arrays that may be used inside another schedule
+            function must appear in this list.
+    Returns:
+        None or a tuple of ndarray: If the function dynamically
+            allocates its output arrays, they must be returned in a tuple
+            of arrays.
+
+    """
+    func_name = None
+    zero_args = False
+    if len(dec_args) == 1 and not dec_kwargs and callable(dec_args[0]):
+        callable_arg = dec_args[0]
+        zero_args = True
+    elif dec_kwargs:
+        if 'func_name' in dec_kwargs:
+            func_name = dec_kwargs['func_name']
+
+    def wrap(func):
+        def wrapped_func(*args, **kwargs):
+            # Save arguments, function, and results pointers/references
+            # to the schedule list:
+            # If trace mode is on, add to schedule.
+            schedule_function = chainer.config.schedule_func
+            if schedule_function is not None:
+                # Note: 'ret = func(*args, **kwargs)' is called inside
+                # the following method.
+                ret = schedule_function.append_function(func, args, kwargs,
+                                                        func_name=func_name)
+                # Add the schedule function as an attribute of the
+                # FunctionNode instance (or more generally, to any class)
+                # that contains the wrapped function as a method
+                if len(args) > 0:
+                    instance = args[0]
+                    if inspect.isclass(instance):
+                        # note: this is not currently needed.
+                        instance.schedule_func = schedule_function
+            else:
+                ret = func(*args, **kwargs)
+            return ret
+        return wrapped_func
+    if zero_args:
+        return wrap(callable_arg)
+    else:
+        return wrap
+
+
+def static_forward_optimizations(func, inputs):
+    """Perform checks needed for creation of a static schedule.
+
+    Check if `func` supports static graph optimizations. If not, try
+    to automatically wrap it to be compatible.
```

Comment:
~~automatically~~

What does it do if it fails to wrap it?

Reply:
Sorry, it was incorrectly worded. It should not fail, but will raise an exception if it does for some reason.